cl2_test.go

```diff
@@ -0,0 +1,230 @@
+		"{\n" +
+			"  \"properties\": {\n" +
+			"    \"FROM_ST\": [ { \"anything-but\": [ \"441\" ] } ]\n" +
+			"  }\n" +
+			"}",
+	}
+	anythingButMatches := []int{211158, 210411, 96682, 120, 210615}
+
+	/* will add when we have numeric
+	complexArraysRules := []string{
+		"{\n" +
+			"  \"geometry\": {\n" +
+			"    \"type\": [ \"Polygon\" ],\n" +
+			"    \"coordinates\": {\n" +
+			"      \"x\": [ { \"numeric\": [ \"=\", -122.42916360922355 ] } ]\n" +
+			"    }\n" +
+			"  }\n" +
+			"}",
+		"{\n" +
+			"  \"geometry\": {\n" +
+			"    \"type\": [ \"MultiPolygon\" ],\n" +
+			"    \"coordinates\": {\n" +
+			"      \"y\": [ { \"numeric\": [ \"=\", 37.729900216217324 ] } ]\n" +
+			"    }\n" +
+			"  }\n" +
+			"}",
+		"{\n" +
+			"  \"geometry\": {\n" +
+			"    \"coordinates\": {\n" +
+			"      \"x\": [ { \"numeric\": [ \"<\", -122.41600944012424 ] } ]\n" +
+			"    }\n" +
+			"  }\n" +
+			"}",
+		"{\n" +
+			"  \"geometry\": {\n" +
+			"    \"coordinates\": {\n" +
+			"      \"x\": [ { \"numeric\": [ \">\", -122.41600944012424 ] } ]\n" +
+			"    }\n" +
+			"  }\n" +
+			"}",
+		"{\n" +
+			"  \"geometry\": {\n" +
+			"    \"coordinates\": {\n" +
+			"      \"x\": [ { \"numeric\": [ \">\",  -122.46471267081272, \"<\", -122.4063085128395 ] } ]\n" +
+			"    }\n" +
+			"  }\n" +
+			"}",
+	bm = newBenchmarker()
+	bm.addRules(complexArraysRules, complexArraysMatches)
+	fmt.Printf("COMPLEX_ARRAYS events/sec: %.1f\n", bm.run(lines, t))
+	}
+	complexArraysMatches := []int{227, 2, 149444, 64368, 127485}
+	*/
+
+	lines := getCL2Lines(t)
+	fmt.Printf("lines: %d\n", len(lines))
+
+	// initial run to stabilize memory
+	bm := newBenchmarker()
+	bm.addRules(exactRules, exactMatches)
+
+	bm.run(t, lines)
+
+	bm = newBenchmarker()
+	bm.addRules(exactRules, exactMatches)
+	fmt.Printf("EXACT events/sec: %.1f\n", bm.run(t, lines))
+
+	bm = newBenchmarker()
+	bm.addRules(anythingButRules, anythingButMatches)
+	fmt.Printf("ANYTHING-BUT events/sec: %.1f\n", bm.run(t, lines))
+}
+
+type benchmarker struct {
+	wanted map[X]int
+	q      *Quamina
+}
+
+func newBenchmarker() *benchmarker {
+	q, _ := New()
+	return &benchmarker{q: q, wanted: make(map[X]int)}
+}
+
+func (bm *benchmarker) addRules(rules []string, wanted []int) {
+	for i, rule := range rules {
+		rname := fmt.Sprintf("r%d", i)
+		_ = bm.q.AddPattern(rname, rule)
+		bm.wanted[rname] = wanted[i]
+	}
+	fmt.Println(matcherStats(bm.q.matcher.(*coreMatcher)))
+}
+
+func (bm *benchmarker) run(t *testing.T, events [][]byte) float64 {
+	t.Helper()
+	gotMatches := make(map[X]int)
+	before := time.Now()
+	for _, event := range events {
+		matches, err := bm.q.MatchesForEvent(event)
+		if err != nil {
+			t.Error("m4e: " + err.Error())
+		}
+		for _, match := range matches {
+			got, ok := gotMatches[match]
+			if !ok {
+				got = 1
+			} else {
+				got++
+			}
+			gotMatches[match] = got
+		}
+	}
+	elapsed := float64(time.Since(before).Milliseconds())
+	eps := float64(cityLotsLineCount) / (elapsed / 1000.0)
```

Comment:
When running only this test, I get 0 events per  second..
```
quamina on  v1.0prep
 ❯ go test -v -run "^TestCl2"
=== RUN   TestCl2
lines: 213068
Field matchers: 6 (avg size 1.000, max 1), Value matchers: 1, SmallTables 44 (avg size 3.128, max 7), singletons 0
Field matchers: 6 (avg size 1.000, max 1), Value matchers: 1, SmallTables 44 (avg size 3.128, max 7), singletons 0
EXACT events/sec: 0.0
Field matchers: 6 (avg size 4.000, max 4), Value matchers: 4, SmallTables 72 (avg size 3.056, max 5), singletons 0
ANYTHING-BUT events/sec: 0.0
--- PASS: TestCl2 (0.96s)
PASS
ok      github.com/timbray/quamina      1.225s
```

This is because - 
```suggestion
	eps := float64(len(events)) / (elapsed / 1000.0)
```

Reply:
Yeah, that was just a bug, already fixed.